### ISSUE TITLE
EDIT: impossible to reload a composition if there is no pattern of type keys

### ISSUE THREAD

**Issue by askoz** [gridsound/daw#52]
Hello,

a very interesting application !

I would like to point out some problems about the new "Drag and Drop raw file" feature

 -  Copyrighted music cannot be streamed
 
 -  "Export to Json" feature does not work in presence of raw files

**Comment by mr21** [comment]
yo :) sorry for the delay

What happens in those scenario, because it's seems to works with short samples (WAV or MP3)?
By copyrighted music do you just mean a buffer of 3/4 minutes long or something special with the fact that it's copyrighted and maybe protected with something?
The samples are not saved into the JSON, only hashes to recognize them when you reload and drop them back.

**Comment by askoz** [comment]
The box "Drag and Drop raw file" works fine with WAV or MP3 but if I export in json, close the application window or use it next day, import is not usable anymore

Copyrighted music problem it's more general

If you place a box that can allow a user to put whatever he wants and let's suppose you will consider to let people publish their compositions through the net, there are some aspects even of security you have to consider

An artist protects the reproduction of his own soundwork, also the most dramatic scenario could be a recorded terroristic message saved in mp3 and diffused from your app ... aaargh !!!

For the moment it's ok in this way, but an architecture design to prevent some problems could be

![SoundStream](https://user-images.githubusercontent.com/22682952/64340549-ba81b080-cfe6-11e9-8e57-01b358678b79.png)

**Comment by mr21** [comment]
Hi, thanks a lot for your interest!

About the technical bug aspect here
> The box "Drag and Drop raw file" works fine with WAV or MP3 but if I export in json, close the application window or use it next day, import is not usable anymore

I don't follow.. i have retested right now and everything work as expected, if you see "missing data" on your samples, this is normal, you have to drop the mp3/wav again into it.

And if you are talking about dropping the `.gs` file then i've retested too and it's working for me... could you open the console dev tools and tell me if you see something unusual?

**Comment by mr21** [comment]
I like your schema :D indeed a report button will be needed.. but for the moment we can't upload any samples on the database, only JSON.
Something to take into account is the users will have to pay in order to upload raw audio files to the gridsound servers. Only JSON can be stocked freely anyway..

About the copyrighted stuff, could i just right in the CGU that users have to comply to their local laws (so fair use, or idk) and delegate the whole thing behing the report button?

**Comment by askoz** [comment]
I recorded what happens in my browser, if you can look at video file in the zip attachment
[ice_video_20190905-175213.zip](https://github.com/gridsound/daw/files/3580162/ice_video_20190905-175213.zip)

**Comment by mr21** [comment]
Well... this is weird :/ is there something in the console?
Does the composition was created with an older version? normally there is no pb but idk

**Comment by askoz** [comment]
Chrome version is 76.0.3809.132 (Build official) (64 bit)

In my office Chrome updates are blocked by administrator due to security policy

In attachment the .gs file used in the import test

[MixComposition.zip](https://github.com/gridsound/daw/files/3583450/MixComposition.zip)

**Comment by mr21** [comment]
well.. i have also the bug, thanks you for your help 

The bug is due to the fact that they are no pattern of type keys in your composition... the DAW try to open the first keys pattern for the pianoroll even if none :/ sorry for this inconvenient

**Comment by mr21** [comment]
``` diff
--- MixComposition.gs	2019-09-06 15:50:40.546412800 +0200
+++ MixComposition-fixed.gs	2019-09-06 15:50:42.970427900 +0200
@@ -13,6 +13,7 @@
 		"2": { "type": "audio/mp3", "duration": 267.299, "hash": "1-004279920-79339d835cac7b93c1dfb4810a1f52b1c6f4e8" }
 	},
 	"patterns": {
+		"0": { "type": "keys", "dest": "1", "keys": "0", "duration": 4, "name": "pat", "order": 0 },
 		"1": { "type": "buffer", "dest": "1", "buffer": "1", "duration": 697, "name": "Kalimba", "order": 0 },
 		"2": { "type": "buffer", "dest": "2", "buffer": "2", "duration": 535, "name": "Kozminski-Il_cinese", "order": 1 }
 	},
@@ -59,5 +60,7 @@
 			}
 		}
 	},
-	"keys": {}
+	"keys": {
+		"0": {}
+	}
 }
```

If you apply this fix, your composition shouldn't have any problem anymore

**Comment by askoz** [comment]
OK, I've applied your patch and now works properly

For the copyright stuff pay attention because in case of success
someone will knock at your door ...

If you are interested I can do an Apache Web Server component that reads your json data
to translate them into audio samples, and send them to a Streaming Server

Bye

**Comment by mr21** [comment]
I didn't forgot what you said, could you create a new issue just for that? thanks a lot anyway

**Comment by askoz** [comment]
OK, I close this issue and make a new one related to copyright

**Comment by mr21** [comment]
Let me reopen it because you have spot a bug too, and this issue has to stay open until the bug is gone, thanks again btw

**Comment by mr21** [comment]
The bug is fixed, we can now load a cmp without any pattern of type `keys` :)
